File: Python/test_person_analysis.py
import unittest
from person_analysis import person_analysis

class test_person_analysis(unittest.TestCase):

# Перевірка коректності розрахунку BMI.
    def test_calculate_bmi(self):
        person = person_analysis("John Doe", 30, 175, 70)
        self.assertAlmostEqual(person.calculate_bmi(), 22.86, places=2)
        
        # Хибний тест AssertionError: 22.857142857142858 != 25.0 within 2 places (2.1428571428571423 difference)
        person = person_analysis("Faulty Person", 30, 175, 70)
        self.assertAlmostEqual(person.calculate_bmi(), 25.00, places=2)  

# Перевірка стану здоров'я "Underweight".
    def test_get_health_status_underweight(self):
        person = person_analysis("Jane Doe", 25, 165, 45)
        self.assertEqual(person.get_health_status(), "Underweight")
        
        # Хибний тест AssertionError: 'Underweight' != 'Normal weight'
        person = person_analysis("Faulty Jane Doe", 25, 165, 45)
        self.assertEqual(person.get_health_status(), "Normal weight")


# Перевірка стану здоров'я "Normal weight".
    def test_get_health_status_normal_weight(self):
        person = person_analysis("John Doe", 30, 175, 70)
        self.assertEqual(person.get_health_status(), "Normal weight")
        
        # Хибний тест AssertionError: 'Normal weight' != 'Overweight'
        person = person_analysis("Faulty John Doe", 30, 175, 70)
        self.assertEqual(person.get_health_status(), "Overweight") 


# Перевірка стану здоров'я "Overweight".
    def test_get_health_status_overweight(self):
        person = person_analysis("Jim Doe", 35, 180, 85)
        self.assertEqual(person.get_health_status(), "Overweight")
        
        # Хибний тест AssertionError: 'Overweight' != 'Obesity'
        person = person_analysis("Faulty Jim Doe", 35, 180, 85)
        self.assertEqual(person.get_health_status(), "Obesity")


# Перевірка стану здоров'я "Obesity".
    def test_get_health_status_obesity(self):
        person = person_analysis("Jake Doe", 40, 170, 100)
        self.assertEqual(person.get_health_status(), "Obesity")
        
        # Хибний тест AssertionError: 'Obesity' != 'Underweight'
        person = person_analysis("Faulty Jake Doe", 40, 170, 100)
        self.assertEqual(person.get_health_status(), "Underweight")


# Перевірка рядкового представлення об'єкта.
    def test_str(self):
        person = person_analysis("John Doe", 30, 175, 70)
        expected_string = ("Name: John Doe, Age: 30, Height: 175, Weight: 70, "
                           "BMI: 22.857142857142858, Health Status: Normal weight")
        self.assertEqual(str(person), expected_string)
        
        # Хибний тест AssertionError: 'Name[44 chars]ght: 70, BMI: 22.857142857142858, Health Status: Normal weight' != 'Name[44 chars]ght: 70, BMI: 25.00, Health Status: Obesity'
        person = person_analysis("Faulty John Doe", 30, 175, 70)
        expected_string = ("Name: Faulty John Doe, Age: 30, Height: 175, Weight: 70, "
                           "BMI: 25.00, Health Status: Obesity") 
        self.assertEqual(str(person), expected_string)


if __name__ == '__main__':
    unittest.main()

File: PHP/test_person_analysis.php
<?php
use PHPUnit\Framework\TestCase;
require_once 'person_analysis.php';

class test_person_analysis extends TestCase {

    // Перевірка коректності розрахунку BMI
    public function testCalculateBMI() {
        $person = new person_analysis("John Doe", 25, 180, 75);
        $this->assertEquals(23.15, round($person->calculateBMI(), 2));
    }

    // Перевірка розрахунків станів здоров'я
    public function testGetHealthStatusNormalWeight() {
        $person = new person_analysis("John Doe", 25, 180, 75);
        $this->assertEquals("Normal weight", $person->getHealthStatus());
    }

    public function testGetHealthStatusUnderweight() {
        $person = new person_analysis("Jane Doe", 22, 160, 45);
        $this->assertEquals("Underweight", $person->getHealthStatus());
    }

    public function testGetHealthStatusOverweight() {
        $person = new person_analysis("Mark Smith", 30, 170, 80);
        $this->assertEquals("Overweight", $person->getHealthStatus());
    }

    public function testGetHealthStatusObesity() {
        $person = new person_analysis("Lucy Brown", 28, 160, 85);
        $this->assertEquals("Obesity", $person->getHealthStatus());
    }

    // Перевірка рядкового представлення об'єкта
    public function testToString() {
        $person = new person_analysis("John Doe", 25, 180, 75);
        $expectedString = "Name: John Doe, Age: 25, Height: 180, Weight: 75, BMI: 23.148148148148, Health Status: Normal weight";
        $this->assertEquals($expectedString, (string)$person);
    }



    // Хибний тест (зріст = 0)
    public function testCalculateBMIInvalidHeight() {
        $this->expectException(InvalidArgumentException::class);
        $person = new person_analysis("John Doe", 25, 0, 75);
        $person->calculateBMI();
    }

    // Хибний тест (вага < 0)
    public function testCalculateBMIInvalidWeight() {
        $this->expectException(InvalidArgumentException::class);
        $person = new person_analysis("John Doe", 25, 180, -75);
        $person->calculateBMI();
    }

    // Хибний тест (відсутнє ім'я)
    public function testMissingName() {
        $this->expectException(InvalidArgumentException::class);
        $person = new person_analysis("", 25, 180, 75);
        $person->__toString();
    }

    // Хибний тест (вік == null)
    public function testMissingAge() {
        $this->expectException(InvalidArgumentException::class);
        $person = new person_analysis("John Doe", null, 180, 75);
        $person->__toString();
    }

    // Хибний тест перевірка рядкового представлення об'єкта (дробове значення BMI округлене)
    public function testMissingToString() {
        $person = new person_analysis("John Doe", 25, 180, 75);
        $expectedString = "Name: John Doe, Age: 25, Height: 180, Weight: 75, BMI: 23.15, Health Status: Normal weight";
        $this->assertEquals($expectedString, (string)$person);
    }
}
?>